### ISSUE TITLE
How to create custom field type that imitates Object (JSON) behavior

### ISSUE THREAD

**Issue by sambhav2612** [keystonejs/keystone#1992]
Using v5 KeystoneJS

Aforementioned, I've come across a scenario which requires me to create an Object of Integer type fields as:

```
    fields: {
        user: {
            type: Relationship,
            ref: 'User'
        },
        parameters: {
           focus: {type: Integer, defaultValue: 0},
           smoothness: {type: Integer, defaultValue: 0},
           legal: {type: Integer, defaultValue: 0},
           braking: {type: Integer, defaultValue: 0},
           distraction: {type: Integer, defaultValue: 0},
           rash_driving: {type: Integer, defaultValue: 0},
        },
    },
```

For now, we are resorting on creating a Relationship between the models (yes, we had to break this object into a whole model of its own!).

Now, I found [this](https://github.com/keystonejs/keystone/issues/1825) issue however since sit only imitates Text behaviour, it didn't work out for us. I've been trying to understand the [official documentation](https://www.keystonejs.com/guides/custom-field-types) as well but can't seem to work it out as well due to lack of any examples online.

Can anyone help?

**Comment by Vultraz** [comment]
So, if I read this right, what you want is to end up with something that looks like?
```js
{
  focus: 0,
  smoothness: 0,
  legal: 0,
  breaking:  0,
  distraction: 0,
  rash_driving: 0,
}
```

**Comment by sambhav2612** [comment]
@Vultraz yes, also, should save in the `parameters` namespace.

**Comment by sambhav2612** [comment]
@optnio00

**Comment by MadeByMike** [comment]
There is no current mechanism for this. You can could create a custom field type. But the details of that depend on each specific implementation details and don't belong here unless they relate to issues or bugs.

Creating a general field type like this is either a duplicate of #315 or #195

**Comment by Vultraz** [comment]
Actually, why do you need a custom field type at all? If you have a list with each of your parameters as fields, like so:
```
    fields: {
        user: { type: Relationship, ref: 'User' },
        focus: {type: Integer, defaultValue: 0},
        smoothness: {type: Integer, defaultValue: 0},
        legal: {type: Integer, defaultValue: 0},
        braking: {type: Integer, defaultValue: 0},
        distraction: {type: Integer, defaultValue: 0},
        rash_driving: {type: Integer, defaultValue: 0},
    },
```

Then you perform a GQL query that includes all of your fields, you'll get a response that can be parsed to JSON and give you the exact object structure you're looking for.

Reposting the same fetch helper from your other issue:
```js
fetch('/admin/api', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      variables,
      query,
    }),
  }).then(result => result.json());
```

**Comment by sambhav2612** [comment]
@Vultraz actually there's more to this model than I wrote, it has more fields and we need distinction between these parameters from all other fields in the model henceforth we went with the 2 level hierarchy

**Comment by molomby** [comment]
I've just seen this issue. It touches on some interesting conversations we've been having lately around encoding standard field types within JSON.

Basically the idea is that, once the JSON DB adapter (#1177) is done, field types will have specific JSON encode/decode functionality baked in. This should let Keystone create "JSON-ready" object from any list item. If we also borrow some machinery from the JSON field type (#315) it opens up some interesting possibilities, like: 

* Storing a subset (or multiple subsets) of lists fields in a JSON blob, rather than regular DB columns
* Nesting fields (as per the original question)
* Storing a "history" of item changes in a JSON array, within the item record
* Importing/exporting to JSON

The first point in particular is interesting I think as it would let us app developers do Mongo-like prototyping of their list schemas, then easily transition to a proper relational schema once their models had settled down. Keystone could do this transparently and maybe by default -- so if, for example, you added a field to a list without adding a column to the relevant table, your values would be stored in a `_otherFields` JSON column until you updated your DB structure.
